src/python_testing/TC_DEM_2_8.py

```diff
@@ -0,0 +1,299 @@
+            TestStep("20a", "TH reads Forecast attribute.",
+                     "Value has to include ForecastUpdateReason=InternalOptimization"),
+            TestStep("21", "TH sends CancelRequest.",
+                     "Verify DUT responds with status INVALID_IN_STATE(0xcb)"),
+            TestStep("22", "TH sends TestEventTrigger command to General Diagnostics Cluster on Endpoint 0 with EnableKey field set to PIXIT.DEM.TEST_EVENT_TRIGGER_KEY and EventTrigger field set to PIXIT.DEM.TEST_EVENT_TRIGGER for Constraints-based Adjustment Adjustment Test Event Clear.",
+                     "Verify DUT responds with status SUCCESS(0x00)"),
+        ]
+
+        return steps
+
+    @async_test_body
+    async def test_TC_DEM_2_8(self):
+        # pylint: disable=too-many-locals, too-many-statements
+        """Run the test steps."""
+        self.step("1")
+        # Commission DUT - already done
+
+        # Subscribe to Events and when they are sent push them to a queue for checking later
+        events_callback = EventChangeCallback(Clusters.DeviceEnergyManagement)
+        await events_callback.start(self.default_controller,
+                                    self.dut_node_id,
+                                    self.matter_test_config.endpoint)
+
+        self.step("2")
+        await self.check_test_event_triggers_enabled()
+
+        self.step("3")
+        await self.send_test_event_trigger_constraint_based_adjustment()
+
+        self.step("3a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kOnline)
+
+        self.step("3b")
+        forecast = await self.read_dem_attribute_expect_success(attribute="Forecast")
+        asserts.assert_is_not_none(forecast.slots[0].manufacturerESAState)
+
+        self.step("3c")
+        await self.check_dem_attribute("OptOutState", Clusters.DeviceEnergyManagement.Enums.OptOutStateEnum.kNoOptOut)
+
+        self.step("4")
+        # Matter UTC is time since 00:00:00 1/1/2000
+        now = datetime.datetime.now(tz=datetime.timezone.utc) - datetime.datetime(2000, 1, 1, 0, 0, 0, 0, datetime.timezone.utc)
+
+        constraintList = [Clusters.DeviceEnergyManagement.Structs.ConstraintsStruct(
+            startTime=now.total_seconds() - 10, duration=20, loadControl=0)]
+        await self.send_request_constraint_based_forecast(constraintList, cause=Clusters.DeviceEnergyManagement.Enums.AdjustmentCauseEnum.kLocalOptimization, expected_status=Status.ConstraintError)
+
+        self.step("5")
+        # Matter UTC is time since 00:00:00 1/1/2000
+        now = datetime.datetime.now(tz=datetime.timezone.utc) - datetime.datetime(2000, 1, 1, 0, 0, 0, 0, datetime.timezone.utc)
+
+        constraintList = [Clusters.DeviceEnergyManagement.Structs.ConstraintsStruct(startTime=now.total_seconds() + 10, duration=20, loadControl=0),
+                          Clusters.DeviceEnergyManagement.Structs.ConstraintsStruct(
```

Comment:
As per TC DEM 2_7 consider making this easier to read by formatting the lines so we just focus on the startTime values

Reply:
Done